.grit/patterns/openai_v4.md

```diff
@@ -11,20 +11,80 @@ engine marzano(0.1)
 language js
 
 // Rewrite the constructor
-`new $constructor($params)` where {
-    $constructor <: `OpenAIApi` => `OpenAI`,
-    $params <: [$config],
-    $program <: contains or {
-        `const $config = new Configuration($details)`,
-        `let $config = new Configuration($details)`,
-        `var $config = new Configuration($details)`
-    } => .,
-    $params => `$details`,
-    $program <: contains `import $old from $src` where {
-      $src <: `"openai"`,
-      $old => `OpenAI`
+
+pattern change_constructor() {
+    `new $constructor($params)` where {
+        $constructor <: `OpenAIApi` => `OpenAI`,
+        $params <: [$config],
+        $program <: contains or {
+            `const $config = new Configuration($details)`,
+            `let $config = new Configuration($details)`,
+            `var $config = new Configuration($details)`
+        } => .,
+        $params => `$details`,
+        $program <: contains `import $old from $src` where {
+        $src <: `"openai"`,
+        $old => `OpenAI`
+        }
+    }
+}
+
+pattern change_chat_completion() {
+    or {
+        member_expression($object, $property) where {
+            $object <: js"openai",
+            $property <: js"createChatCompletion" => js"chat.completions.create"
+        },
+        js"chatCompletion.data.choices" => js"chatCompletion.choices"
```

Comment:
Unfortunately this part is too brittle. It assumes the `chatCompletion` is the name of the previous call. It actually can be anything. What we really want to do is split this pattern in two:

1. Find (and replace) the createChatCompletion
2. For every case where we *reference* that chatCompletion, rewrite the code accessing it.